docs/proposals/01-firmware-update.md

```diff
@@ -0,0 +1,381 @@
+- the solution SHOULD allow automated hardware servers' firmware lifecycle maintaining;
+- the solution MUST be extensible by the possibility of using plugins for update strategy;
+- the solution MUST be extensible by the possibility of adding vendor-specific update jobs;
+- the solution SHOULD be as kubernetes-native as possible;
+
+### Non-Goals
+
+## Proposal
+
+### Custom resources
+
+The following CRs aimed to represent the current state of a particular server, a desired state of a group of servers and available firmware versions for a particular manufacturer-model:
+
+- [ServerFirmware](#serverfirmware)
+- [ServerFirmwareGroup](#serverfirmwaregroup)
+- [DiscoveredFirmware](#discoveredfirmware)
+
+All the following CRs MUST be cluster-scoped.
+The firmware versions defined for concrete `ServerFirmware` object MUST take precedence over those provided by corresponding `ServerFirmwareGroup` object.
+
+#### ServerFirmware
+
+`ServerFirmware` CR represents the desired state of concrete hardware server.
+The `.spec` of this type references the `Server` object, reflects its `.status.bios` field and contains the list of firmwares desired to be installed.
+The `.status` of this type contains information about the BIOS/firmware versions which are actually installed on the server.
+Aside from that `.spec` contains the scan threshold and the `.status` contains last scan operation timestamp.
+These two fields allow calculating whether the scanning for installed firmware is required or not.
+The `ServerFirmware` object SHOULD be created along with corresponding `Server` object and MUST be unique across the cluster.
+
+```yaml
+apiVersion: metal.ironcore.dev/v1alpha1
+kind: ServerFirmware
+metadata:
+  name: foo
+spec:
+  scanThreshold: 30m
+  serverRef:
+    name: foo
+  bios:
+    version: 1.0.0
+  firmwares:
+    - name: ssd
+      manufacturer: ACME Corp.
+      version: 1.0.0
+    - name: nic
+      manufacturer: Intel
+      version: 2.0.0
+status:
+  lastScanTime: 01-01-2001 01:00:00
+  bios:
+    version: 1.0.0
+  firmwares:
+    - name: ssd
+      manufacturer: ACME Corp.
+      version: 1.0.0
+    - name: nic
+      manufacturer: Intel
+      version: 2.0.0
+```
+
+#### ServerFirmwareGroup
+
+`ServerFirmwareGroup` CR represents the groups of hardware servers and their desired firmware versions.
+The group of servers MUST be manufacturer- and model-specific to ensure that defined firmware will be applicable.
+The `.spec` of this type contains built-in label selector and the list of firmwares desired to be installed.
+The `.status` of this type contains the information about number of servers within the defined group which are in desired state.
+The group MUST have a unique selector and selectors in different `ServerFirmwareGroup` objects MUST NOT intersect.
+
+```yaml
+apiVersion: metal.ironcore.dev/v1alpha1
+kind: ServerFirmwareGroup
+metadata:
+  name: bar-group
+spec:
+  manufacturer: Lenovo
+  model: 7x21
```

Comment:
How are Servers identified by model? The current api only mentions the [manufacturer](https://github.com/ironcore-dev/metal-operator/blob/main/api/v1alpha1/server_types.go#L146).

Reply:
Unfortunately, I do not have answer. Regarding Lenovo, we can get server's model from SKU. What about other manufacturers - no idea yet. Maybe some API exists from which it could be retrieved.
Thoughts behind the necessity of knowing the server's model is that different models could use different hardware and potentially different firmware versions. Hence to be able to automate the process we need to somehow deal with it.